File: 159 Longest Substring with At Most Two Distinct Characters.js
"use strict";
/*
Longest Substring with At Most Two Distinct Characters
https://leetcode.com/problems/longest-substring-with-at-most-two-distinct-characters/
*/
var lengthOfLongestSubstringTwoDistinct = function(input){
    var length = input.length;
    if(length < 3) return length;
    var res = '';
    var i,j,c,v;
    var map = new Map();
    for(i=0, j=0;i<length;i++){
        c = input.charAt(i);
        v = map.get(c);
        if(v === undefined){
            map.set(c,1);
        } else {
            map.set(c,v+1);
        }
        while(map.size > 2){
            c = input.charAt(j);
            v = map.get(c);
            if(v === 1){
                map.delete(c);
            } else {
                map.set(c,v-1);
            }
            j++;
        }
        if(i+1-j > res.length) res = input.slice(j,i+1);
    }
    console.log(res);
    return res.length;
}

var input = "abaerwfacsqefav";
console.log(lengthOfLongestSubstringTwoDistinct(input));

File: 159 Longest Substring with At Most Two Distinct Characters.py
'''
159. Longest Substring with At Most Two Distinct Characters

Given a string s , find the length of the longest substring t that contains at most 2 distinct characters.

Example 1:

Input: "eceba"
Output: 3
Explanation: t is "ece" which its length is 3.

Example 2:

Input: "ccaabbb"
Output: 5
Explanation: t is "aabbb" which its length is 5.
'''
from collections import Counter

class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        # Same method as in 340. Longest Substring with At Most K Distinct Characters
        k = 2
        
        if len(s) <= k:
            return len(s)
        
        # Result
        maxSize = k
        
        # Sliding window caches
        left = 0
        book = Counter()
        distinctCount = 0
        
        for i, c in enumerate(s):
            book[c] += 1
            if book[c] == 1:
                distinctCount += 1
                
            # Shrink
            while distinctCount > k:
                lc = s[left]
                book[lc] -= 1
                if book[lc] == 0:
                    distinctCount -= 1
                left += 1
   
            maxSize = max(maxSize, i - left + 1)
                    
        return maxSize
